src/core/thread/key_manager.cpp

```diff
@@ -243,6 +244,22 @@ Error KeyManager::SetNetworkKey(const NetworkKey &aKey)
         child.SetMleFrameCounter(0);
     }
 #endif
+}
+
+Error KeyManager::SetNetworkKey(const NetworkKey &aKey)
+{
+    Error error = kErrorNone;
+
+    SuccessOrExit(Get<Notifier>().Update(mNetworkKey, aKey, kEventNetworkKeyChanged));
```

Comment:
Right, I was doing this in dataset before the last update and following the changes to dataset, this needs updating.

How about we do a `GetNetworkKey()` instead of using `mNetworkKey`, in `Update`. This way `GetNetworkKey() `will always return `NetworkKey`, irrespective of the configuration used (for PSA it reads from ITS for non-PSA returns mNetworkKey). Then we just do the same Update followed by Store.

So the code will now change to:

```
Error KeyManager::SetNetworkKey(const NetworkKey &aKey)
{
    Error error = kErrorNone;

    SuccessOrExit(Get<Notifier>().Update(GetNetworkKey(), aKey, kEventNetworkKeyChanged));
    Get<Notifier>().Signal(kEventThreadKeySeqCounterChanged);

#if OPENTHREAD_CONFIG_PLATFORM_KEY_REFERENCES_ENABLE
    IgnoreError(StoreNetworkKey(true));
#endif

    mKeySequence = 0;
    UpdateKeyMaterial();
    ResetFrameCounters();

exit:
    return error;
}
```